pkg/reconciler/managed/reconciler.go

```diff
@@ -701,12 +718,17 @@ func NewReconciler(m manager.Manager, of resource.ManagedKind, o ...ReconcilerOp
 		supportedManagementPolicies: defaultSupportedManagementPolicies(),
 		log:                         logging.NewNopLogger(),
 		record:                      event.NewNopRecorder(),
+		metricRecorder:              NewNopMetricRecorder(),
+		stateRecorder:               statemetrics.NewNopStateRecorder(),
 	}
 
 	for _, ro := range o {
 		ro(r)
 	}
 
+	// State recorder is started in the background to record MR states.
+	go r.stateRecorder.Run(context.Background(), schema.GroupVersionKind(of))
```

Comment:
Never leak go routines from a constructor. This must be started outside with an available`ctx`.

Reply:
I made state recorders `Runnable` and registered to manager to manage go routine's lifecycle with it.